Use presence of _symint in kernel name to generate symint sig or not

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #84579

Something people found confusing was that whether or not a native::
signature would get SymInt or not in its type was based on the dispatch
key.  This changes it so that SymInt or not in type is based on whether
or not you have _symint in the name of the kernel or not.  This means
that even when we make operators support SymInt, you no longer have to
go and update all the preexisting definitions; instead, you now
selectively write _symint to opt individual kernels into SymInt support.

I then go and update a bunch of kernels that don't have proper SymInt
support to make use of this convention.  There is some hacking around
for view generation code.

I also add support for external backends to specify 'symint' operators, for which we generate SymInt signatures instead of regular signatures.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D39310060](https://our.internmc.facebook.com/intern/diff/D39310060)

ezyang committed:
Use presence of _symint in kernel name to generate symint sig or not

Something people found confusing was that whether or not a native::
signature would get SymInt or not in its type was based on the dispatch
key.  This changes it so that SymInt or not in type is based on whether
or not you have _symint in the name of the kernel or not.  This means
that even when we make operators support SymInt, you no longer have to
go and update all the preexisting definitions; instead, you now
selectively write _symint to opt individual kernels into SymInt support.

I then go and update a bunch of kernels that don't have proper SymInt
support to make use of this convention.  There is some hacking around
for view generation code.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Use presence of _symint in kernel name to generate symint sig or not"


Something people found confusing was that whether or not a native::
signature would get SymInt or not in its type was based on the dispatch
key.  This changes it so that SymInt or not in type is based on whether
or not you have _symint in the name of the kernel or not.  This means
that even when we make operators support SymInt, you no longer have to
go and update all the preexisting definitions; instead, you now
selectively write _symint to opt individual kernels into SymInt support.

I then go and update a bunch of kernels that don't have proper SymInt
support to make use of this convention.  There is some hacking around
for view generation code.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Use presence of _symint in kernel name to generate symint sig or not"


Something people found confusing was that whether or not a native::
signature would get SymInt or not in its type was based on the dispatch
key.  This changes it so that SymInt or not in type is based on whether
or not you have _symint in the name of the kernel or not.  This means
that even when we make operators support SymInt, you no longer have to
go and update all the preexisting definitions; instead, you now
selectively write _symint to opt individual kernels into SymInt support.

I then go and update a bunch of kernels that don't have proper SymInt
support to make use of this convention.  There is some hacking around
for view generation code.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Use presence of _symint in kernel name to generate symint sig or not"


Something people found confusing was that whether or not a native::
signature would get SymInt or not in its type was based on the dispatch
key.  This changes it so that SymInt or not in type is based on whether
or not you have _symint in the name of the kernel or not.  This means
that even when we make operators support SymInt, you no longer have to
go and update all the preexisting definitions; instead, you now
selectively write _symint to opt individual kernels into SymInt support.

I then go and update a bunch of kernels that don't have proper SymInt
support to make use of this convention.  There is some hacking around
for view generation code.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Use presence of _symint in kernel name to generate symint sig or not"


Something people found confusing was that whether or not a native::
signature would get SymInt or not in its type was based on the dispatch
key.  This changes it so that SymInt or not in type is based on whether
or not you have _symint in the name of the kernel or not.  This means
that even when we make operators support SymInt, you no longer have to
go and update all the preexisting definitions; instead, you now
selectively write _symint to opt individual kernels into SymInt support.

I then go and update a bunch of kernels that don't have proper SymInt
support to make use of this convention.  There is some hacking around
for view generation code.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Use presence of _symint in kernel name to generate symint sig or not"


Something people found confusing was that whether or not a native::
signature would get SymInt or not in its type was based on the dispatch
key.  This changes it so that SymInt or not in type is based on whether
or not you have _symint in the name of the kernel or not.  This means
that even when we make operators support SymInt, you no longer have to
go and update all the preexisting definitions; instead, you now
selectively write _symint to opt individual kernels into SymInt support.

I then go and update a bunch of kernels that don't have proper SymInt
support to make use of this convention.  There is some hacking around
for view generation code.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

wconstab commented:
> Something people found confusing was that whether or not a native::
signature would get SymInt or not in its type was based on the dispatch
key

Before this PR, was the idea that 'meta' kernels were getting c++ symint signatures but real kernels weren't, and now this becomes more explicit because for either meta or composite kernels you just have to write _symint in the name and take out the guesswork?

ezyang commented:
> Before this PR, was the idea that 'meta' kernels were getting c++ symint signatures but real kernels weren't, and now this becomes more explicit because for either meta or composite kernels you just have to write _symint in the name and take out the guesswork?

Yep

wconstab commented on torchgen/api/types.py:
```diff
@@ -783,10 +783,12 @@ def kernel_signature(
-        return NativeSignature(f.func, prefix=prefix, symint=backend_index.symint)
+        meta = backend_index.get_kernel(f)
+        return NativeSignature(
+            f.func, prefix=prefix, symint=meta is not None and meta.supports_symint()
```
does having a python decomp count here in favor of being symint compatible, or is a 'meta' kernel always strictly necessary?

ezyang replied:
Python decomp suffices, meta is not required

wconstab commented on torchgen/gen_functionalization_type.py:
```diff
@@ -84,21 +84,21 @@ def gen_composite_view_copy_kernel(g: NativeFunctionsViewGroup) -> Optional[str]
+    return self.reshape(size);
   } else {
-    auto output = at::_ops::view::call(self, size);
+    auto output = at::_ops::view::call(self, c10::SymIntArrayRef::fromIntArrayRef(size));
```
ok what's going on in this kernel? the `view_copy` name above didn't get `_symint` tacked on, and some of the symint related stuff got removed, but then we are constructing symints and invoking view.

wconstab commented on torchgen/native_function_generation.py:
```diff
@@ -303,7 +303,7 @@ def generate_function(
         if func.kind() == SchemaKind.out
         else cpp.name(func)
-    )
+    ) + "_symint"  # TODO: hmm
```
why is _symint assumed here? shouldn't this be driven by the yaml?

ezyang replied:
Sooo I am not entirely sure here, but the idea is that these kernels are autogenerated, and they handle SymInts, so we should give them symint names so they get the right signature. It's a bit dodgy though... I got here by just tweaking until it compiled

wconstab replied:
that sounds reasonable

clarification: these 'generated' kernels are going to call into handwritten kernels (adapting from functional to out or vice versa) right?  if so, would we be potentially having a situation like this?
foo_out <-- handwritten, no symint support
foo_symint <-- generated, just trivially unwraps symints and expects ints, calls foo_out

if so, i still don't think that's a problem.  other than, we should probably make it clearer what's going on

ezyang replied:
Err, not really? The generated kernels typically call into other API as defined by the codegen; e.g., a generated out will generate a call to functional and then copy_ (but the functional call is not a direct native:: call; it'll just be to some dispatched api).

wconstab reviewed (approved):
This sounds like a better policy to me, and it looks correct outside of the minor questions I raised

ezyang commented:
I'm going to squish this with the earlier diff as it won't be possible to get XLA working without both sides of the coin.

ezyang committed:
Update on "Use presence of _symint in kernel name to generate symint sig or not"


Something people found confusing was that whether or not a native::
signature would get SymInt or not in its type was based on the dispatch
key.  This changes it so that SymInt or not in type is based on whether
or not you have _symint in the name of the kernel or not.  This means
that even when we make operators support SymInt, you no longer have to
go and update all the preexisting definitions; instead, you now
selectively write _symint to opt individual kernels into SymInt support.

I then go and update a bunch of kernels that don't have proper SymInt
support to make use of this convention.  There is some hacking around
for view generation code.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Use presence of _symint in kernel name to generate symint sig or not"


Something people found confusing was that whether or not a native::
signature would get SymInt or not in its type was based on the dispatch
key.  This changes it so that SymInt or not in type is based on whether
or not you have _symint in the name of the kernel or not.  This means
that even when we make operators support SymInt, you no longer have to
go and update all the preexisting definitions; instead, you now
selectively write _symint to opt individual kernels into SymInt support.

I then go and update a bunch of kernels that don't have proper SymInt
support to make use of this convention.  There is some hacking around
for view generation code.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Use presence of _symint in kernel name to generate symint sig or not"


Something people found confusing was that whether or not a native::
signature would get SymInt or not in its type was based on the dispatch
key.  This changes it so that SymInt or not in type is based on whether
or not you have _symint in the name of the kernel or not.  This means
that even when we make operators support SymInt, you no longer have to
go and update all the preexisting definitions; instead, you now
selectively write _symint to opt individual kernels into SymInt support.

I then go and update a bunch of kernels that don't have proper SymInt
support to make use of this convention.  There is some hacking around
for view generation code.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Use presence of _symint in kernel name to generate symint sig or not"


Something people found confusing was that whether or not a native::
signature would get SymInt or not in its type was based on the dispatch
key.  This changes it so that SymInt or not in type is based on whether
or not you have _symint in the name of the kernel or not.  This means
that even when we make operators support SymInt, you no longer have to
go and update all the preexisting definitions; instead, you now
selectively write _symint to opt individual kernels into SymInt support.

I then go and update a bunch of kernels that don't have proper SymInt
support to make use of this convention.  There is some hacking around
for view generation code.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang commented on torch/csrc/jit/runtime/static/generated_ops.cpp:
```diff
@@ -1812,7 +1812,7 @@ REGISTER_OPERATOR_FUNCTOR(aten::prod, aten_prod, [](Node* n) -> SROperator {
       auto& out = p_node->Output(0).toTensor();
       fastResizeToZero(out);
-      at::native::prod_out(self, dtype, out);
+      at::native::prod_out_symint(self, dtype, out);
```
This diff can be reverted if I only `_symint` autogen functions if the source schema has symint

ezyang commented on torchgen/api/types.py:
```diff
@@ -778,15 +780,18 @@ def kernel_signature(
+    if symint:
+        assert f.func.has_symint(), \
+            f"attempted to define symint kernel for {backend_index.dispatch_key} without SymInt in schema"
+    """
```
Ditto here, less sloppy autogen will let this be restored

ezyang commented on torchgen/gen.py:
```diff
@@ -606,7 +612,7 @@ def __call__(self, f: NativeFunction) -> str:
                     )
```
static dispatch looks like it needs to be updated

ezyang committed:
Update on "Use presence of _symint in kernel name to generate symint sig or not"


Something people found confusing was that whether or not a native::
signature would get SymInt or not in its type was based on the dispatch
key.  This changes it so that SymInt or not in type is based on whether
or not you have _symint in the name of the kernel or not.  This means
that even when we make operators support SymInt, you no longer have to
go and update all the preexisting definitions; instead, you now
selectively write _symint to opt individual kernels into SymInt support.

I then go and update a bunch of kernels that don't have proper SymInt
support to make use of this convention.  There is some hacking around
for view generation code.

For enhance backwards compatibility, I also introduce `_symint` overload to `at::_ops` interface; since operators can be called at two dispatcher signatures (one with SymInt, and one without.)

I also add support for external backends to specify 'symint' operators, for which we generate SymInt signatures instead of regular signatures.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Use presence of _symint in kernel name to generate symint sig or not"


Something people found confusing was that whether or not a native::
signature would get SymInt or not in its type was based on the dispatch
key.  This changes it so that SymInt or not in type is based on whether
or not you have _symint in the name of the kernel or not.  This means
that even when we make operators support SymInt, you no longer have to
go and update all the preexisting definitions; instead, you now
selectively write _symint to opt individual kernels into SymInt support.

I then go and update a bunch of kernels that don't have proper SymInt
support to make use of this convention.  There is some hacking around
for view generation code.

For enhance backwards compatibility, I also introduce `_symint` overload to `at::_ops` interface; since operators can be called at two dispatcher signatures (one with SymInt, and one without.)

I also add support for external backends to specify 'symint' operators, for which we generate SymInt signatures instead of regular signatures.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang commented:
@ezyang has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D39310060).

ezyang committed:
Update on "Use presence of _symint in kernel name to generate symint sig or not"


Something people found confusing was that whether or not a native::
signature would get SymInt or not in its type was based on the dispatch
key.  This changes it so that SymInt or not in type is based on whether
or not you have _symint in the name of the kernel or not.  This means
that even when we make operators support SymInt, you no longer have to
go and update all the preexisting definitions; instead, you now
selectively write _symint to opt individual kernels into SymInt support.

I then go and update a bunch of kernels that don't have proper SymInt
support to make use of this convention.  There is some hacking around
for view generation code.

I also add support for external backends to specify 'symint' operators, for which we generate SymInt signatures instead of regular signatures.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

Differential Revision: [D39310060](https://our.internmc.facebook.com/intern/diff/D39310060)

[ghstack-poisoned]

ezyang commented:
@ezyang has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D39310060).

izaitsevfb commented:
@pytorchbot rebase

seemethere committed:
Update on "Use presence of _symint in kernel name to generate symint sig or not"


Something people found confusing was that whether or not a native::
signature would get SymInt or not in its type was based on the dispatch
key.  This changes it so that SymInt or not in type is based on whether
or not you have _symint in the name of the kernel or not.  This means
that even when we make operators support SymInt, you no longer have to
go and update all the preexisting definitions; instead, you now
selectively write _symint to opt individual kernels into SymInt support.

I then go and update a bunch of kernels that don't have proper SymInt
support to make use of this convention.  There is some hacking around
for view generation code.

I also add support for external backends to specify 'symint' operators, for which we generate SymInt signatures instead of regular signatures.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

Differential Revision: [D39310060](https://our.internmc.facebook.com/intern/diff/D39310060)

[ghstack-poisoned]

seemethere commented:
Rebased to include new xla hash from @zou3519, `f00dd2f35ecf6455d97237d63c70c9c8ec190940`

seemethere commented:
@pytorchbot merge -f "already passed CI, merged internally, need to get in ASAP"
